packages/backend-function/src/factory.ts

```diff
@@ -182,10 +185,26 @@ class FunctionFactory implements ConstructFactory<AmplifyFunction> {
     outputStorageStrategy,
     resourceNameValidator,
   }: ConstructFactoryGetInstanceProps): AmplifyFunction => {
+    // TODO this should be conditional on some signal,
+    // for example that any access to data was granted for this function.
     if (!this.generator) {
+      const dataResources = constructContainer
+        .getConstructFactory<AmplifyData>('DataResources')
```

Comment:
Passing the generic AmplifyData into this correctly types the instance returned, but I don't believe we get any guarantees from this. I am inclined to type narrow with a type guard instead, but wanted input before I introduce that.

Reply:
This is fine. This typing is for us and we know what `DataResources` are. No need to double check that.

But this should be `.getConstructFactory<ResourceProvider<DataResources>>('DataResources')` not full construct class.

Reply:
I made a fix to this. A bit unclear on how this object is meant to be structured and I've made a comment on my [new type](https://github.com/aws-amplify/amplify-backend/pull/2147/files#r1821316875) to continue this conversation